pkg/kv/kvserver/loqrecovery/apply.go

```diff
@@ -365,6 +366,18 @@ func MaybeApplyPendingRecoveryPlan(
 	}
 
 	applyPlan := func(nodeID roachpb.NodeID, plan loqrecoverypb.ReplicaUpdatePlan) error {
+		binaryVersion := clusterversion.ByKey(clusterversion.BinaryVersionKey)
+		binaryMinSupportedVersion := clusterversion.ByKey(clusterversion.BinaryMinSupportedVersionKey)
+		clusterVersion, err := kvstorage.SynthesizeClusterVersionFromEngines(
+			ctx, engines, binaryVersion, binaryMinSupportedVersion,
+		)
+		if err != nil {
+			return errors.Wrap(err, "failed to get cluster version from storage")
+		}
+		if err := checkPlanVersionIsAllowed(plan.Version, clusterVersion.Version); err != nil {
```

Comment:
We definitely need a bypass for stuck upgrade. But wouldn't env var be error prone? It needs to be enabled globally because all nodes need to process the plan regardless of replica placement. Subsequently, once recovery is performed, it has to be turned off. This is identical to running offline recovery, which is not good. Maybe we can have a field on the staging request to update plan with internal version during staging? That would allow cli to stage a plan that would succeed application upon restart.

Reply:
Makes sense, thanks.

Reply:
LGTM.